### PR TITLE
[MI-2300]:Created /confluence migrate start command to start subscription migration

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**

### CODE REVIEW

File: server/command.go [hunk 1]
```diff
@@ -60,6 +60,9 @@ const (
 	configAPIEndpoint      = "%s/api/v4/actions/dialogs/open"
 	configModalTitle       = "Confluence Config"
 	configPerPage          = 10
+	NoOldSubscriptionMsg   = "No old Subscriptions found for migration"
+	MigrationStartMsg      = "The migration process has been completed. Please refer to server logs for more information."
```
Comment (ayusht2810) [inline server/command.go]: shouldn't this be `MigrationCompletedMsg` according to the message?

File: server/command.go [hunk 2]
```diff
@@ -424,6 +431,27 @@ func listOldSubscriptions(p *Plugin, context *model.CommandArgs, args ...string)
 	return &model.CommandResponse{}
 }
 
+func startSubscriptionMigration(p *Plugin, context *model.CommandArgs, args ...string) *model.CommandResponse {
+	oldSubscriptions, gErr := service.GetOldSubscriptions()
```
Comment (ayusht2810) [inline server/command.go]:
```suggestion
	oldSubscriptions, getErr := service.GetOldSubscriptions()
```

File: server/save_subscription.go [hunk 3]
```diff
@@ -26,70 +26,83 @@ func (p *Plugin) handleSaveSubscription(w http.ResponseWriter, r *http.Request)
...
+	instance, err := p.getInstanceFromURL(path)
 	if err != nil {
-		p.LogAndRespondError(w, http.StatusInternalServerError, "Error in loading connection.", err)
-		return
+		return http.StatusInternalServerError, "Not able to get instance from url", err
 	}
 
-	client, err := instance.GetClient(conn)
+	if err = p.HasPermissionToManageSubscription(instance.GetURL(), userID, channelID); err != nil {
+		return http.StatusForbidden, "Don't have the permission to create subscription. Please contact your administrator", err
```
Comment (ayusht2810) [inline server/save_subscription.go]:
```suggestion
		return http.StatusForbidden, "You don't have permission to create a subscription. Please contact your administrator", err
```

File: server/serializer/space_subscription.go [hunk 4]
```diff
@@ -74,6 +74,14 @@ func (ss SpaceSubscription) GetUserID() string {
 	return ss.UserID
 }
 
+func (ss SpaceSubscription) GetEvents() []string {
+	return ss.Events
+}
+
+func (ss SpaceSubscription) GetSpaceKeyOrPageID() string {
```
Comment (ayusht2810) [inline server/serializer/space_subscription.go]: shouldn't this name be just `GetSpaceKey` if we are only returning SpaceKey? same for page subscription function

File: server/save_subscription.go [hunk 5]
```diff
@@ -98,26 +111,13 @@ func (p *Plugin) handleSaveSubscription(w http.ResponseWriter, r *http.Request)
...
-		Message:   subscriptionSaveSuccess,
-	}
-	_ = p.API.SendEphemeralPost(userID, post)
-	ReturnStatusOK(w)
-}
-
-func (p *Plugin) LogAndRespondError(w http.ResponseWriter, statusCode int, errorLog string, err error) {
-	p.API.LogError(errorLog, "Error", err.Error())
-	http.Error(w, errorLog, statusCode)
+	return http.StatusOK, subscriptionSaveSuccess, nil
```
Comment (ayusht2810) [inline server/save_subscription.go]:
```suggestion
	return statusCode, subscriptionSaveSuccess, nil
```

File: server/serializer/space_subscription.go (hunk 4 above)
Comment (Kshitij-Katiyar) [reply server/serializer/space_subscription.go]: actually this function is a part of an interface and this is implemented by 2 structs page and space subscription and for page we have pageID and for space struct we have spaceKey

**Review by ayusht2810 — APPROVED** [review]

File: server/command.go [hunk 6]
```diff
@@ -60,6 +60,9 @@ const (
 	configAPIEndpoint      = "%s/api/v4/actions/dialogs/open"
 	configModalTitle       = "Confluence Config"
 	configPerPage          = 10
+	NoOldSubscriptionMsg   = "No old Subscriptions found for migration"
```
Comment (raghavaggarwal2308) [inline server/command.go]:
```suggestion
	NoOldSubscriptionsMsg   = "No old Subscriptions found for migration."
```

File: server/save_subscription.go [hunk 7]
```diff
@@ -26,70 +26,83 @@ func (p *Plugin) handleSaveSubscription(w http.ResponseWriter, r *http.Request)
...
+	instance, err := p.getInstanceFromURL(path)
 	if err != nil {
-		p.LogAndRespondError(w, http.StatusInternalServerError, "Error in loading connection.", err)
-		return
+		return http.StatusInternalServerError, "Not able to get instance from url", err
 	}
 
-	client, err := instance.GetClient(conn)
+	if err = p.HasPermissionToManageSubscription(instance.GetURL(), userID, channelID); err != nil {
+		return http.StatusForbidden, "You don't have permission to create a subscription. Please contact your administrator.", err
```
Comment (raghavaggarwal2308) [inline server/save_subscription.go]:
```suggestion
		return http.StatusForbidden, "You don't have the permission to create a subscription. Please contact your administrator.", err
```

File: server/save_subscription.go [hunk 8]
```diff
@@ -26,70 +26,83 @@ func (p *Plugin) handleSaveSubscription(w http.ResponseWriter, r *http.Request)
 	subscriptionType := params[ParamSubscriptionType]
 	userID := r.Header.Get(config.HeaderMattermostUserID)
 
-	instance, err := p.getInstanceFromURL(r.URL.Path)
+	body, err := ioutil.ReadAll(r.Body)
 	if err != nil {
-		http.Error(w, err.Error(), http.StatusInternalServerError)
+		p.LogAndRespondError(w, http.StatusInternalServerError, "", err)
```
Comment (raghavaggarwal2308) [inline server/save_subscription.go]: Shouldn't we add a errorLog message here instead of empty string

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

File: go.mod [hunk 9]
```diff
@@ -7,10 +7,10 @@ require (
 	github.com/gorilla/mux v1.8.0
 	github.com/levigross/exp-html v0.0.0-20120902181939-8df60c69a8f5
 	github.com/mattermost/mattermost-plugin-api v0.0.26-0.20220223141232-cb8b1984774a
-	github.com/mattermost/mattermost-server/v6 v6.5.0
+	github.com/mattermost/mattermost-server/v6 v6.7.2
```
Comment (manojmalik20) [inline go.mod:10]: What was the reason for this change?

File: server/command.go [hunk 10]
```diff
@@ -60,6 +60,9 @@ const (
 	configAPIEndpoint      = "%s/api/v4/actions/dialogs/open"
 	configModalTitle       = "Confluence Config"
 	configPerPage          = 10
+	NoOldSubscriptionsMsg  = "No old Subscriptions found for migration."
+	MigrationCompletedMsg  = "The migration process has been completed. Please refer to server logs for more information."
```
Comment (manojmalik20) [inline server/command.go]:
```suggestion
	MigrationCompletedMsg  = "The migration process has been completed. Please refer to the server logs for more information."
```

File: server/serializer/page_subscription.go [hunk 11]
```diff
@@ -75,6 +75,14 @@ func (ps PageSubscription) GetConfluenceURL() string {
 	return ps.GetSubscription().BaseURL
 }
 
+func (ps PageSubscription) GetEvents() []string {
```
Comment (manojmalik20) [inline server/serializer/page_subscription.go]:
```suggestion
func (ps *PageSubscription) GetEvents() []string {
```
Why aren't we using this here?

File: server/subscription_migration.go [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	SubscriptionType string   `json:"subscriptionType"`
+	Alias            string   `json:"alias"`
+	BaseURL          string   `json:"baseURL"`
+	SpaceKey         string   `json:"spaceKey"`
+	PageID           string   `json:"pageID"`
+	ChannelID        string   `json:"channelID"`
+	Events           []string `json:"events"`
+}
+
+func (p *Plugin) migrateSubscription(subscriptions []serializer.Subscription, userID string) string {
```
Comment (manojmalik20) [inline server/subscription_migration.go]:
```suggestion
func (p *Plugin) migrateSubscriptions(subscriptions []*serializer.Subscription, userID string) string {
```

File: server/subscription_migration.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+			Alias:            sub.GetAlias(),
+			BaseURL:          sub.GetConfluenceURL(),
+			SpaceKey:         sub.GetSpaceKeyOrPageID(),
+			PageID:           sub.GetSpaceKeyOrPageID(),
+			ChannelID:        sub.GetChannelID(),
+			Events:           sub.GetEvents(),
+		})
+		if err != nil {
+			failedSubscriptionLog = err.Error()
+			p.API.LogError("Unable to marshal request body for subscription", "Subscription", sub, "Error", err.Error())
```
Comment (manojmalik20) [inline server/subscription_migration.go:41]: Shouldn't there be a `continue` statement after this line?
You should append the error in `failedSubscription` variable and then `continue` after this line.

File: server/subscription_migration.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+			failedSubscription = fmt.Sprintf("%s- %s:%s\n", failedSubscription, sub.GetAlias(), failedSubscriptionLog)
+		} else {
+			subscriptionCreated = fmt.Sprintf("%s- %s\n", subscriptionCreated, sub.GetAlias())
+		}
+	}
+	if subscriptionCreated == subscriptionCreatedHeader {
+		subscriptionCreated = ""
+	}
+	if failedSubscription == subscriptionFailedHeader {
+		subscriptionCreated = ""
```
Comment (manojmalik20) [inline server/subscription_migration.go]:
```suggestion
		failedSubscription = ""
```

File: go.mod (hunk 9 above)
Comment (Kshitij-Katiyar) [reply go.mod:10]: took the pull from the latest master from matter most.This came from there

File: server/serializer/page_subscription.go (hunk 11 above)
Comment (Kshitij-Katiyar) [reply server/serializer/page_subscription.go]: all the previous was like this that's why done like this. Should I leave this as it is or do this in every function.?

Comment (manojmalik20) [reply server/serializer/page_subscription.go]: I think you should do this in every function.

File: server/subscription_migration.go [hunk 15]
```diff
@@ -24,11 +24,10 @@ type MigrateSubscriptionBody struct {
 	Events           []string `json:"events"`
 }
 
-func (p *Plugin) migrateSubscription(subscriptions []serializer.Subscription, userID string) string {
+func (p *Plugin) migrateSubscriptions(subscriptions []serializer.Subscription, userID string) string {
```
Comment (manojmalik20) [inline server/subscription_migration.go:27]:
```suggestion
func (p *Plugin) migrateSubscriptions(subscriptions []*serializer.Subscription, userID string) string {
```

**Review by manojmalik20 — APPROVED** [review]

File: server/subscription_migration.go (hunk 15 above)
Comment (manojmalik20) [reply server/subscription_migration.go:27]: You didn't do this change and still resolved the conversation. Why?

**Pull request merged**